src/test/java/com/shapesecurity/bandolier/BundlerTest.java

```diff
@@ -142,63 +150,118 @@ public void testDeterminism() throws Exception {
 		modules.put("/js4.js", "export var d = 'd'");
 		TestLoader localLoader = new TestLoader(modules);
 
-		String actual = TestUtils.toString(TestUtils.bundleStandard("/js1.js", resolver, localLoader));
-
-		String expected = "(function(global){\n" +
-			"\"use strict\";\n" +
-			"function require(file,parentModule){\n" +
-			"if({\n" +
-			"}.hasOwnProperty.call(require.cache,file))return require.cache[file];\n" +
-			"var resolved=require.resolve(file);\n" +
-			"if(!resolved)throw new Error(\"Failed to resolve module \"+file);\n" +
-			"var module$={\n" +
-			"id:file,require:require,filename:file,exports:{\n" +
-			"},loaded:false,parent:parentModule,children:[]};\n" +
-			"if(parentModule)parentModule.children.push(module$);\n" +
-			"var dirname=file.slice(0,file.lastIndexOf(\"/\")+1);\n" +
-			"require.cache[file]=module$.exports;\n" +
-			"resolved.call(void 0,module$,module$.exports,dirname,file);\n" +
-			"module$.loaded=true;\n" +
-			"return require.cache[file]=module$.exports;\n" +
-			"}require.modules={\n" +
-			"};\n" +
-			"require.cache={\n" +
-			"};\n" +
-			"require.resolve=function(file){\n" +
-			"return{\n" +
-			"}.hasOwnProperty.call(require.modules,file)?require.modules[file]:void 0;\n" +
-			"};\n" +
-			"require.define=function(file,fn){\n" +
-			"require.modules[file]=fn;\n" +
-			"};\n" +
-			"require.define(\"1\",function(module,exports,__dirname,__filename){\n" +
-			"var __resolver=require(\"2\",module);\n" +
-			"var b=__resolver[\"b\"];\n" +
-			"var __resolver=require(\"3\",module);\n" +
-			"var c=__resolver[\"c\"];\n" +
-			"var result=\"a\"+b+c;\n" +
-			"exports[\"result\"]=result;\n" +
-			"});\n" +
-			"require.define(\"2\",function(module,exports,__dirname,__filename){\n" +
-			"var __resolver=require(\"4\",module);\n" +
-			"var d=__resolver[\"d\"];\n" +
-			"var b=\"b\"+d;\n" +
-			"exports[\"b\"]=b;\n" +
-			"});\n" +
-			"require.define(\"3\",function(module,exports,__dirname,__filename){\n" +
-			"var __resolver=require(\"4\",module);\n" +
-			"var d=__resolver[\"d\"];\n" +
-			"var c=\"c\"+d;\n" +
-			"exports[\"c\"]=c;\n" +
-			"});\n" +
-			"require.define(\"4\",function(module,exports,__dirname,__filename){\n" +
-			"var d=\"d\";\n" +
-			"exports[\"d\"]=d;\n" +
-			"});\n" +
-			"return require(\"1\");\n" +
-			"}.call(this,this).result);\n";
-
-		assertEquals(expected, actual);
+		String actualPierced = TestUtils.toString(TestUtils.bundlePierced(BundlerOptions.SPEC_OPTIONS, "/js1.js", resolver, localLoader));
+		String actualStandard = TestUtils.toString(TestUtils.bundleStandard(BundlerOptions.SPEC_OPTIONS, "/js1.js", resolver, localLoader));
+
+		String expectedPierced = "(function(t){\n" +
```

Comment:
Is there any way to the nice short output a previous version of this PR had? I know it's not correct in edge cases, but sometimes that's an acceptable tradeoff given the overhead here.

The short output being something like
```js
(function(){
  "use strict";
  var d = "d";
  var c = "c" + d;
  var b = "b" + d;
  var result = "a" + b + c;
  return { result: result };
})();
```

Reply:
Added a `dangerLevel` options, an enum which can be `SAFE` (previous operation), `BALANCED` (pretty much what you posted for example, but asserts imports are read-only), and `DANGEROUS` (what we had before, and exactly what you posted for example, with the object being predeclared before returning).

Reply:
Nice! Do you think it would be possible to statically identify some subset of programs where a theoretically dangerous level is in fact safe? For example, if there's no cycles, no `import *`, and no attempt to write to an imported binding, does that make it the `BALANCED` level actually 100% safe? If so, a good followup (not necessarily part of this PR) would be to automatically choose a more efficient strategy when it's safe to do so.

Reply:
If there are no cycles/self-imports, it's safe to assume there are no TDZ issues, however `BALANCED` also currently removes the Symbol.toStringTag entry on the module full-namespace import. This would require no occurrences of `import * as X` (as opposed to `import *`), or require us to assert that `X[Symbol.toStringTag]` is never accessed, which can be done in some cases (say only `StaticMemberExpression`s were the only accesses of the namespace reference ).

`DANGEROUS` mode could be identified by enabling it and having a different read-detection response. For example, if there is a possible read, throw a bundle-time error rather than inserting a lambda with a `TypeError` throw statement.

Reply:
>  would require no occurrences of `import * as X`

Sorry, yeah, that's what I meant.

> say only `StaticMemberExpressions` were the only accesses of the namespace reference

And exported module does not export any (non-arrow) functions which refer to `this` or `eval`, which can also give you a reference to the namespace object. Probably this analysis is more trouble than it's worth, but the basic "there are no namespace exports" one seems doable.

> For example, if there is a possible read

Do you mean a possible write, rather than a possible read?

---

I guess any such analysis also needs the assumption that each exported binding is written to exactly once, right? (Or I guess, to be precise, that there are no writes to exported bindings inside of closures; multiple writes at the top level would be acceptable.) Otherwise the whole "live binding" complication comes into play. Still seems doable though.

Reply:
Yes, I mean't possible write. The exported binding can be written as much as desired from the exporting class, however, the imported binding in that module or others is unwritable (as opposed to immutable, as any changes from the underlying module will be immediately reflected in the imported binding).